File: DSASheet/3_SearchingAndSorting/2/Product array puzzle/Solution.js
class Solution {
    productExceptSelf(nums, n){
      //code here
      var arr_1 = Array(n).fill(1);
      var arr_2 = Array(n).fill(1);
      
      for(var i = 1; i < nums.length; i ++){
        arr_1[i] = nums[i - 1] * arr_1[i - 1];
        var lastIndex = n - i - 1;
        arr_2[lastIndex] = arr_2[lastIndex + 1] * nums[lastIndex + 1]
        
      }
      var result = [];
      for(var i = 0; i < nums.length; i ++){
          result[i] = arr_1[i] * arr_2[i]
      }
      return result
    }
}

File: DSASheet/3_SearchingAndSorting/2/Product array puzzle/Solution.py
class Solution:
    def productExceptSelf(self, arr, n):
        #code here
        arr_1 = [0] * n
        arr_2 = [0] * n
        arr_1[0] = arr_2[-1] = 1
        result = []
        print(arr_1, arr_2)
        for i in range(1, n):
            arr_1[i] = arr[i - 1] * arr_1[i - 1]
            last_index = n - i - 1
            arr_2[last_index] = arr_2[last_index + 1] * arr[last_index + 1]
            
        return [arr_1[i] * arr_2[i] for i in range(n)]
    
arr = [10, 3, 5, 6, 2]  
arr = [0, 19]  
print(Solution().productExceptSelf(arr, len(arr)))